farmerbot/server/farmerbot.go

```diff
@@ -0,0 +1,296 @@
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/big"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/constants"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/manager"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/models"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/parser"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/version"
+	"github.com/threefoldtech/tfgrid-sdk-go/rmb-sdk-go/peer"
+)
+
+// FarmerBot for managing farms
+type FarmerBot struct {
+	configFile   string
+	config       *models.Config
+	sub          substrate.Substrate
+	powerManager *manager.PowerManager
+	dataManager  manager.DataManager
+	twinID       uint32
+	network      string
+	mnemonic     string
+	identity     substrate.Identity
+}
+
+// NewFarmerBot generates a new farmer bot
+func NewFarmerBot(ctx context.Context, configFile, network, mnemonic string) (FarmerBot, error) {
+	farmerbot := FarmerBot{mnemonic: mnemonic, network: network, configFile: configFile}
+
+	substrateManager := substrate.NewManager(constants.SubstrateURLs[network]...)
+	sub, err := substrateManager.Substrate()
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.sub = *sub
+
+	err = farmerbot.setConfig()
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	// TODO: check farm hex seed if available
+	identity, err := substrate.NewIdentityFromSr25519Phrase(mnemonic)
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.identity = identity
+
+	twinID, err := sub.GetTwinByPubKey(identity.PublicKey())
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.twinID = twinID
+
+	powerManager := manager.NewPowerManager(identity, sub, farmerbot.config)
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.powerManager = &powerManager
+
+	rmb, err := peer.NewRpcClient(ctx, peer.KeyTypeSr25519, mnemonic, constants.RelayURLS[network], fmt.Sprintf("farmerbot-rpc-%d", farmerbot.config.Farm.ID), sub, true)
+	if err != nil {
+		return FarmerBot{}, errors.Wrap(err, "could not create rmb client")
+	}
+
+	rmbNodeClient := manager.NewRmbNodeClient(rmb)
+	dataManager := manager.NewDataManager(sub, farmerbot.config, &rmbNodeClient)
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.dataManager = dataManager
+
+	return farmerbot, nil
+}
+
+// Run runs farmerbot to update nodes and power management
+func (f *FarmerBot) Run(ctx context.Context) {
+	if err := f.start(ctx); err != nil {
+		log.Fatal().Err(err).Msg("[FARMERBOT] error starting")
+	}
+
+	go f.update(ctx)
+	go f.serve(ctx)
+
+	log.Info().Msg("[FARMERBOT] up and running...")
+
+	// graceful shutdown
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+	<-sigChan
+
+	shutdownCtx, shutdownRelease := context.WithTimeout(context.Background(), 10*time.Second)
+	defer shutdownRelease()
```

Comment:
This is absurd, Why the farmer bot care how or when the context would be cancelled. please there is a HUGE different between library and cmdline application.

The farmer bot takes a context, all what it needs to do is graceful shutdown (basically return from the main loop) if the ctx was cancelled it shoud not care how the context is canelled.

The user of the FarmerBot object then can decide when and how the context will be cancelled. The command line can then listen on signals, os something else, and cancel the context accordingly